public/js/repeater.js

```diff
@@ -0,0 +1,185 @@
+window.Repeater = (function() {
+  "use strict"
+
+  var repeatedSets
+
+  // Since we can't depend on Array.prototype.map
+  function map(arr, fn) {
+    var out = new Array(arr.length)
+    arr.forEach(function(x, i) { out[i] = fn(x, i) })
+    return out
+  }
+
+  function indexBy(key, arr) {
+    var out = {}
+    arr.forEach(function(x) { out[x[key]] = x })
+    return out
+  }
+
+  function Block() { this.init.apply(this, arguments) }
+  (function(_) {
+    _.init = function(el) {
+      var self = this // closure fix
+
+      this.name = el.id
+      this.container = el
+
+      this.instances = map(el.querySelectorAll('.repeater-instance'), function(fieldset, i) {
+        var instance = new Instance(self, fieldset, i)
+        instance.reindex(i)
+        return instance
+      })
+    }
+
+    _.setupListeners = function() {
+      // we use one global listener for removal, because the repeat links may
+      // get added and removed from the DOM, and this means we don't have to
+      // re-register the event listeners when we clone the nodes.
+      this.container.addEventListener('click', function(evt) {
+        if (!evt.target.classList.contains('remove-repeat-link')) return
+
+        evt.preventDefault()
+        var instance = instanceFor(evt.target)
+        instance && instance.remove()
+      })
+
+      return this
+    }
+
+    _.repeat = function() {
+      if (!this.instances.length) throw new Error('empty instances, can\'t repeat!')
+
+      var newIndex = this.instances.length
+      var newEl = this.instances[0].el.cloneNode(true)
+      var newInstance = new Instance(this, newEl, newIndex)
+      newInstance.reindex(newIndex, true)
+      this.container.appendChild(newEl)
+      this.instances.push(newInstance)
+      return newInstance
+    }
+  })(Block.prototype)
+
+  // one instance of a repeater
+  function Instance() { return this.init.apply(this, arguments) }
+  (function(_) {
+    // private functions
+    function reindex(str, index) {
+      // it's always going to be the first [0] or [1] or etc.
+      return str.replace(/\[\d+\]/, '['+index+']')
```

Comment:
Again here `[${index}])` would prefer template literals

Reply:
oh! I should say, this is front-end javascript, run on the browser, so I'm not sure what the compatibility story is for these kinds of features. Can we rely on template literals in our users' browsers?

Reply:
No worries Webpack will take care of this to make it backwards compat to ES5  or whatever we want to target.